### PR TITLE
typechecker: Check jakt `main` type declaration

### PR DESCRIPTION
**Pull request by Pythonen**
Add separate typecheck for main functions as stated in https://github.com/SerenityOS/jakt/issues/184
Also fix tests to use this new declaration.

### CODE REVIEW

**Review by linusg — CHANGES_REQUESTED** [review]
>  Also fix tests to use this new mandatory declaration.

I never said they should be mandatory...

**Comment by Pythonen** [comment]
> > Also fix tests to use this new mandatory declaration.
> 
> I never said they should be mandatory...

Sorry, now I'm a bit lost. So the type declaration for main should be inferred if not explicitly typed by the user or?

**Comment by linusg** [comment]
Right, I think inheriting a bit of magic from C/C++ isn't a bad thing here. Personally I think it should be fine to:

- Omit the return type declaration and not return anything -> implicit `return 0`
- Omit the return type declaration and return a number
- Omit the args parameter if they are not being used

In other words, 

```jakt
function main() {}
```

is the same as

```jakt
function main(args: [String]) throws -> c_int {}
```

but that does not mean that

```jakt
function main(foo: i32, bar: i32) -> String {}
```

should be accepted.

**Comment by Pythonen** [comment]
Right, my bad. Misunderstood it.

File: src/typechecker.rs [hunk 1]
```diff
@@ -1844,6 +1847,53 @@ fn typecheck_function(
...
+        "Main function must take a single array of strings as its parameter".to_string(),
+        function.name_span,
+    ));
+
+    if function.params.len() > 1 {
+        return param_type_error;
+    }
+
+    if !function.params.is_empty() {
+        match &function.params[0].variable.ty {
```
Comment (linusg) [inline src/typechecker.rs:1861]: Just noticed I missed this in my `ty` removal & renamed it now, please rebase/update

File: src/typechecker.rs [hunk 2]
```diff
@@ -1844,6 +1847,53 @@ fn typecheck_function(
...
+                        return None;
+                    }
+                }
+                _ => {
+                    return param_type_error;
+                }
+            },
+            _ => {
+                return param_type_error;
+            }
```
Comment (linusg) [inline src/typechecker.rs:1877]:
```suggestion
            ParsedType::Array(ty, _) => match &**ty {
                ParsedType::Empty => {}
                ParsedType::Name(ty, _) if ty == "String" => {}
                _ => return param_type_error,
            },
            _ => return param_type_error,
```

File: src/typechecker.rs [hunk 3]
```diff
@@ -1844,6 +1847,53 @@ fn typecheck_function(
...
+        function.name_span,
+    ));
+    match &function.return_type {
+        ParsedType::Empty => {}
+        ParsedType::Name(name, _) => {
+            if name != "c_int" {
+                return return_type_error;
+            }
+        }
+        _ => return return_type_error,
```
Comment (linusg) [inline src/typechecker.rs:1892]:
```suggestion
        ParsedType::Empty => {}
        ParsedType::Name(name, _) if name == "c_int" => {}
        _ => return return_type_error,
```

**Review by linusg — COMMENTED** [review]
Looks much better, thanks! Please rebase, and squash the commits while you're at it :)
For each given commit in time, all tests should be passing, so fixing one after the fact is not desirable and makes bisecting harder.

File: src/typechecker.rs (hunk 1 above)
Comment (Pythonen) [reply src/typechecker.rs:1879]: Is it okay though to use it in 
`ParsedType::Array(ty, _) => match &**ty`
or should these destructurings also have more verbose names?

Comment (linusg) [reply src/typechecker.rs:1879]: it's not ok to use anywhere :)

**Review by linusg — CHANGES_REQUESTED** [review]
Typo: `programs'` -> `program's`, and please clean up the commit message, it now contains all the messages from the squashed commits

**Comment by Pythonen** [comment]
Sorry for the hassle, my git-fo is a bit rusty (heh)

**Pull request merged**